### PR TITLE
migrate crate-details, match_version, db:: and test-fakes to sqlx

### PR DESCRIPTION
**Pull request by syphar**
Finding a working smaller piece to migrate became harder and harder, this is bigger than I hoped, but works. 

This migrates some major parts to sqlx & async, also including parts used in our main `web::rustdoc` endpoints. 

Namely: 
- `CrateDetails` and `match_version`, and the remaining release-details / list endpoints
- `db::add_package` methods, `db::file`
- `test::fakes` is async internally with sync wrapper methods for old tests 

For things I didn't want to migrate (yet) I had to sprinkle plenty of `runtime.block_on` around to make it work. 

I think there is a nicer way to design the whole `TestEnvironment` / `FakeRelease` story, perhaps with having async versions of both classes, similar to `AsyncStorage` / `Storage`. I would leave fixing 

Also includes the changes from #2324

### CODE REVIEW

**Comment by syphar** [comment]
@rust-lang/docs-rs-reviewers any chance for a review? 

I know it's much code, but I would love to push this forward. After these async refactors I could then invest some time into making our web handlers much more readable / understandable.

**Comment by GuillaumeGomez** [comment]
I can try but my review won't be up to much as I don't much this part of the code.

**Review by GuillaumeGomez — APPROVED** [review]
Apart from the strange JSON files, this is just a crate change so nothing to note in particular from me.

File: .sqlx/query-007a6d355f893370b544dda0d6129096357aa95c67c76905a6709a66247450ff.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply .sqlx/query-007a6d355f893370b544dda0d6129096357aa95c67c76905a6709a66247450ff.json:1]: Is it really useful?

File: src/db/add_package.rs [hunk 2]
```diff
@@ -92,69 +92,71 @@ pub(crate) fn add_package_into_database(
...
-    conn: &mut Client,
+pub(crate) async fn add_doc_coverage(
+    conn: &mut sqlx::PgConnection,
     release_id: i32,
     doc_coverage: DocCoverage,
 ) -> Result<i32> {
     debug!("Adding doc coverage into database");
-    let rows = conn.query(
+    Ok(sqlx::query_scalar!(
         "INSERT INTO doc_coverage (
```
[the remark this answers is not in the document]
Comment (syphar) [reply src/db/add_package.rs:160]: this is coming from `sqlx` for compile time checked sql queries. 

Locally the queries will be checked against your local database, when built in CI / for docker, it's using the JSON files. 

Initial PR was #2223

**Comment by syphar** [comment]
> Apart from the strange JSON files, this is just a crate change so nothing to note in particular from me.

It's a little more than a crate change, since many methods get converted from sync to async. 

But generally yes ;)

**Comment by GuillaumeGomez** [comment]
Yes absolutely, but it didn't seems strange or anything. So please just add a readme file in the json folder so people like me will know what's going on in there and why it even exists in the first place. :)

**Pull request merged**